src/BloomBrowserUI/bookEdit/bloomField/bloomField.ts

```diff
@@ -0,0 +1,254 @@
+            range.selectNodeContents($(field).find('p').first()[0]);
+        }
+        else{
+            range.selectNodeContents($(field).find('p').last()[0]);
+        }
+        range.collapse(position === CursorPosition.start);//true puts it at the start
+        var sel = window.getSelection();
+        sel.removeAllRanges();
+        sel.addRange(range);
+    }
+    static ManageWhatHappensIfTheyDeleteEverything(field: HTMLElement) {
+        // if the user types (ctrl+a, del) then we get an empty element or '<br></br>', and need to get a <p> in there.
+        // if the user types (ctrl+a, 'blah'), then we get
+
+        $(field).on("input", function (e) {
+            if ($(this).find('p').length === 0) {
+                BloomField.ModifyForParagraphMode(this);
+
+                // Now put the cursor in the paragraph, *after* the character they may have just typed or the
+                // text they just pasted.
+                BloomField.MoveCursorToFirstParagraph(field, CursorPosition.end);
+            }
+        });
+    }
+    // Some custom templates have image containers embedded in bloom-editable divs, so that the text can wrap
+    // around the picture. The problems is that the user can do (ctrl+a, del) to start over on the text, and
+    // inadvertently remove the embedded images. So we introduced the "bloom-preventRemoval" class, and this
+    // tries to safeguard element bearing that class.
+
+    static PreventRemovalOfSomeElements(field:HTMLElement) {
+        var numberThatShouldBeThere = $(field).find(".bloom-preventRemoval").length;
+        if(numberThatShouldBeThere > 0) {
+            $(field).on("input", function (e) {
+                if ($(this).find(".bloom-preventRemoval").length < numberThatShouldBeThere) {
+                    document.execCommand('undo');
+                    e.preventDefault();
+                }
+            });
+        }
+
+        //OK, now what if the above fails in some scenario? This adds a last-resort way of getting
+        //bloom-editable back to the state it was in when the page was first created, by having
+        //the user type in RESETRESET and then clicking out of the field.
+        $(field).blur(function (e) {
+            if ($(this).html().indexOf('RESETRESET') > -1) {
+                $(this).remove();
+                alert("Now go to another book, then back to this book and page.");
+            }
+        });
+    }
+
+    // Work around a bug in geckofx. The effect was that if you clicked in a completely empty text box
+    // the cursor is oddly positioned and typing does nothing. There is evidence that what is going on is that the focus
+    // is on the English qtip (in the FF inspector, the qtip block highlights when you type). https://jira.sil.org/browse/BL-786
+    // This bug mentions the cursor being in the wrong place: https://bugzilla.mozilla.org/show_bug.cgi?id=904846
+    static PrepareNonParagraphField(field:HTMLElement) {
```

Comment:
This seems to be doing the same thing as one branch of HandleFieldFocus, though with a much more comprehensive explanation for why. Can we fix the duplication?